### PR TITLE
[CI] Show sdk size update on PR

### PR DESCRIPTION
**Pull request by testableapple**
### 🔗 Issue Links

Resolve https://stream-io.atlassian.net/browse/PBE-4314

### 🎯 Goal

Show SDKs' size update on PR

### 📺 Showcase

Screenshot 2024-07-24 at 5 56 13 PM

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] Changelog is updated with new localization keys
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

gif

### CODE REVIEW

**Review by laevandus — APPROVED** [review]

**Review by laevandus — DISMISSED** [review]
Approved too quick, CI was failing

File: fastlane/Fastfile [hunk 1]
```diff
@@ -13,12 +13,18 @@ github_repo = ENV['GITHUB_REPOSITORY'] || 'GetStream/stream-chat-swift'
 stress_tests_cycles = 50
 derived_data_path = 'derived_data'
 source_packages_path = 'spm_cache'
-performance_path = "performance/#{github_repo.split('/').last}.json"
+performance_git = 'git@github.com:GetStream/stream-swift-performance-benchmarks.git'
+xcmetrics_path = "performance/#{github_repo.split('/').last}-xcmetrics.json"
```
Comment (testableapple) [inline fastlane/Fastfile:17]: FYI: I'll rename this file in the private repo as soon as this PR is merged.

File: fastlane/Fastfile [hunk 2]
```diff
@@ -854,36 +844,82 @@ desc 'Show current frameworks size'
...
-  if is_ci && !ENV['GITHUB_PR_NUM'].to_s.empty?
-    markdown_table = "## SDK Size\n| `StreamChat` | `StreamChatUI` | `Total` |\n| - | - | - |\n"
-    markdown_table << "|#{options[:sizes][:stream_chat]}MB|#{options[:sizes][:stream_chat_ui]}MB|#{options[:sizes][:total]}MB|"
-    sh("gh pr comment #{ENV.fetch('GITHUB_PR_NUM')} -b '#{markdown_table}'")
+  sh("git clone #{performance_git} #{File.dirname(sdk_size_path)}")
+  performance_benchmarks = JSON.parse(File.read(sdk_size_path))
+  expected_performance = performance_benchmarks['develop']
+  actual_performance = options[:sizes] || frameworks_sizes
+
+  markdown_table = "## SDK Size\n| `title` | `benchmark` | `branch` | `performance` | `status` |\n| - | - | - | - | - |\n"
```
Comment (nuno-vieira) [inline fastlane/Fastfile:854]: Maybe better to name this diff? Performance in this scenario does not make much sense 🤔 
```suggestion
  markdown_table = "## SDK Size\n| `title` | `benchmark` | `branch` | `diff` | `status` |\n| - | - | - | - | - |\n"
```

File: StreamChat.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -13985,10 +13984,10 @@
 			buildSettings = {
 				ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES = YES;
 				CLANG_CXX_LANGUAGE_STANDARD = "gnu++17";
-				CODE_SIGN_STYLE = Manual;
+				CODE_SIGN_IDENTITY = "Apple Development";
+				CODE_SIGN_STYLE = Automatic;
```
Comment (nuno-vieira) [inline StreamChat.xcodeproj/project.pbxproj:13988]: Why are we changing this?

Comment (testableapple) [reply StreamChat.xcodeproj/project.pbxproj:13988]: The cert got expired, I updated it and now it fails for some reason to sign the Test app. So I set it to automatic signing, seems to be working fine now 🤔

File: fastlane/Fastfile (hunk 2 above)
Comment (testableapple) [reply fastlane/Fastfile:854]: What do you think about removing this column at all? 🤔

File: fastlane/Fastfile [hunk 4]
```diff
@@ -854,36 +844,82 @@ desc 'Show current frameworks size'
...
+  ['performance/'].each { |dir| FileUtils.remove_dir(dir, force: true) }
 
-  if is_ci && !ENV['GITHUB_PR_NUM'].to_s.empty?
-    markdown_table = "## SDK Size\n| `StreamChat` | `StreamChatUI` | `Total` |\n| - | - | - |\n"
-    markdown_table << "|#{options[:sizes][:stream_chat]}MB|#{options[:sizes][:stream_chat_ui]}MB|#{options[:sizes][:total]}MB|"
-    sh("gh pr comment #{ENV.fetch('GITHUB_PR_NUM')} -b '#{markdown_table}'")
+  sh("git clone #{performance_git} #{File.dirname(sdk_size_path)}")
+  performance_benchmarks = JSON.parse(File.read(sdk_size_path))
+  expected_performance = performance_benchmarks['develop']
+  actual_performance = options[:sizes] || frameworks_sizes
```
Comment (nuno-vieira) [inline fastlane/Fastfile:852]: Why are these called performances? 🤔

Comment (nuno-vieira) [inline fastlane/Fastfile:852]: At least the last 2 properties should be renamed to something more clear 🤔 
```suggestion
 // What this one is about?
  performance_benchmarks = JSON.parse(File.read(sdk_size_path))
  current_size = performance_benchmarks['develop']
  new_size = options[:sizes] || frameworks_sizes
```

Comment (testableapple) [reply fastlane/Fastfile:852]: Egh, yeah, I was not sure about naming. Just wanted to reuse as much as possible with xcmetrics, so the benchmark would live in https://github.com/GetStream/stream-swift-performance-benchmarks

Maybe it's worth to rename this repo to `apple-internal-benchmarks` and rename those variables in the Fastfile, WDYT?

File: fastlane/Fastfile (hunk 2 above)
Comment (nuno-vieira) [reply fastlane/Fastfile:854]: I think it is useful if we just provide the difference, like "+0.5MB" or "-0.5MB" . The arrows in this case is not very useful

File: fastlane/Fastfile (hunk 4 above)
Comment (testableapple) [reply fastlane/Fastfile]: Updated 👍

File: fastlane/Fastfile (hunk 2 above)
Comment (testableapple) [reply fastlane/Fastfile]: Updated 👍

File: fastlane/Fastfile [hunk 5]
```diff
@@ -854,36 +854,75 @@ desc 'Show current frameworks size'
...
+    status_emoji =
+      if is_total
+        ''
+      elsif diff < 0
+        outstanding_status
+      elsif diff > max_tolerance
+        fail_status
+      elsif diff > fine_tolerance
+        warning_status
+      else
```
Comment (nuno-vieira) [inline fastlane/Fastfile:883]:
```suggestion
      elsif diff < 0
        outstanding_status
      elsif diff >= max_tolerance
        fail_status
      elsif diff >= fine_tolerance
        warning_status
      else
```

We do want to include those values actually. So 0.5 is 🔴  and 0.25 is ⚠️

File: fastlane/Fastfile [hunk 6]
```diff
@@ -854,36 +854,75 @@ desc 'Show current frameworks size'
 lane :show_frameworks_sizes do |options|
   next unless is_check_required(sources: sources_matrix[:size], force_check: @force_check)
 
-  options[:sizes] ||= frameworks_sizes
+  ['metrics/'].each { |dir| FileUtils.remove_dir(dir, force: true) }
+
+  sh("git clone #{metrics_git} #{File.dirname(sdk_size_path)}")
+  sdk_size_benchmarks = JSON.parse(File.read(sdk_size_path))
+  expected_sizes = sdk_size_benchmarks['develop']
```
Comment (nuno-vieira) [inline fastlane/Fastfile:861]: How is this getting the SDK size from develop? Is it returning the current badge value from the README.md?

File: fastlane/Fastfile [hunk 7]
```diff
@@ -854,36 +854,75 @@ desc 'Show current frameworks size'
...
+
+    Dir.chdir(File.dirname(sdk_size_path)) do
+      if sh('git status -s', log: false).to_s.empty?
+        pr_comment_required = false
+        UI.important('No changes in SDK sizes benchmarks. Skipping commit and comment.')
+      else
+        sh('git add -A')
+        sh("git commit -m 'Update #{sdk_size_path}: #{current_branch}'")
+        sh('git push')
+      end
```
Comment (nuno-vieira) [inline fastlane/Fastfile:906]: When is this run? 🤔

**Review by nuno-vieira — COMMENTED** [review]
It is looking great 👌 Just added a couple more questions and suggestions 👍

File: fastlane/Fastfile (hunk 6 above)
Comment (testableapple) [reply fastlane/Fastfile]: From here https://github.com/GetStream/apple-internal-metrics/blob/main/stream-chat-swift-size.json

File: fastlane/Fastfile (hunk 7 above)
Comment (testableapple) [reply fastlane/Fastfile:906]: Every time on CI:
- on PR
- nightly on develop

**Review by nuno-vieira — APPROVED** [review]
LGTM! ✅

**Review by nuno-vieira — APPROVED** [review]

**Pull request merged**